### PR TITLE
Add support for DifferentialTemperatureLoads

### LINKED ISSUE #167: Add support for BarDifferentialTemperatureLoad

### PR DESCRIPTION
**Pull request by JosefTaylor**
### Issues addressed by this PR

Closes #167

### Test files

https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/SAP2000_Toolkit/%23242-DifferentialTemperatureLoad/BuroHappold_BHoM_v4.0.beta_Create%20read%20loads.gh?csf=1&web=1&e=ENijgz

### CODE REVIEW

File: SAP2000_Adapter/CRUD/Read/Loads.cs [hunk 1]
```diff
@@ -520,18 +522,82 @@ private List<ILoad> ReadBarUniformTemperatureLoad(List<string> id = null)
...
+                    if (loadType[i] == 1)
+                    {
+                        loads.Add(new BarUniformTemperatureLoad()
+                        {
+                            TemperatureChange = tempChange,
+                            Loadcase = bhomCases[caseNames[i]],
+                            Objects = new BHoMGroup<Bar> { Elements = { bhomBar } },
+                            Axis = LoadAxis.Global,
+                            Projected = false,
+                        });
```
Comment (peterjamesnugent) [inline SAP2000_Adapter/CRUD/Read/Loads.cs:534]: Worth using the Engine method to benefit from all the checks?

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check installer

**Comment by JosefTaylor** [comment]
@BHoMBot  check required

File: SAP2000_Adapter/CRUD/Create/Loads.cs [hunk 2]
```diff
@@ -591,7 +591,48 @@ private bool CreateLoad(BarPrestressLoad bhLoad)
...
+
+        private bool CreateLoad(BarDifferentialTemperatureLoad bhLoad)
+        {
+            List<Bar> bars = bhLoad.Objects.Elements.ToList();
+            string loadPat = GetAdapterId<string>(bhLoad.Loadcase);
+            string patternName = "None";
+            bool replace = SAPPushConfig.ReplaceLoads;
+            int myType = 1;
+
+            double tempUniform = bhLoad.GetUniformComponent();
```
Comment (peterjamesnugent) [inline SAP2000_Adapter/CRUD/Create/Loads.cs:605]: Does SAP2000 also apply the temperature gradient at the midpoint of the Bar and you have to offset it with a uniform temperature?

File: SAP2000_Adapter/CRUD/Read/Loads.cs [hunk 3]
```diff
@@ -520,18 +526,75 @@ private List<ILoad> ReadBarUniformTemperatureLoad(List<string> id = null)
...
+                                double width = bhomBar.SectionProperty.Vpy + bhomBar.SectionProperty.Vy;
+                                loads.Add(new BarDifferentialTemperatureLoad()
+                                {
+                                    TemperatureProfile = new Dictionary<double, double>() { { 0, 0 }, { 1, val[i] * width } },
+                                    LoadDirection = DifferentialTemperatureLoadDirection.LocalY,
+                                    Loadcase = bhomCases[caseNames[i]],
+                                    Objects = new BHoMGroup<Bar> { Elements = { bhomBar } },
+                                    Axis = LoadAxis.Global,
+                                    Projected = false,
+                                });
```
Comment (peterjamesnugent) [inline SAP2000_Adapter/CRUD/Read/Loads.cs:579]: Was there not an earlier commit that made use of the Engine methods?

File: SAP2000_Adapter/CRUD/Read/Loads.cs [hunk 4]
```diff
@@ -520,18 +526,75 @@ private List<ILoad> ReadBarUniformTemperatureLoad(List<string> id = null)
...
+                            case 3:
+                                double height = bhomBar.SectionProperty.Vpz + bhomBar.SectionProperty.Vz;
+                                loads.Add(new BarDifferentialTemperatureLoad()
+                                {
+                                    TemperatureProfile = new Dictionary<double, double>() { { 0, 0 }, { 1, val[i] * height } },
+                                    LoadDirection = DifferentialTemperatureLoadDirection.LocalZ,
+                                        Loadcase = bhomCases[caseNames[i]],
+                                        Objects = new BHoMGroup<Bar> { Elements = { bhomBar } },
+                                        Axis = LoadAxis.Global,
+                                        Projected = false,
```
Comment (peterjamesnugent) [inline SAP2000_Adapter/CRUD/Read/Loads.cs:590]: As above

**Review by peterjamesnugent — COMMENTED** [review]
Just confirming a few parts of the code - can't approve/comment on the functionality as I don't have SAP2000 installed.

File: SAP2000_Adapter/CRUD/Create/Loads.cs (hunk 2 above)
Comment (JosefTaylor) [reply SAP2000_Adapter/CRUD/Create/Loads.cs:605]: SAP2000 only allows constant gradients, so it has to start at 0, whereas the differential load in BHoM allows any profile.

File: SAP2000_Adapter/CRUD/Read/Loads.cs (hunk 3 above)
Comment (JosefTaylor) [reply SAP2000_Adapter/CRUD/Read/Loads.cs:579]: hmm did I do something weird? lemme check.

Comment (JosefTaylor) [reply SAP2000_Adapter/CRUD/Read/Loads.cs]: ah, took it out of one method, not out of this. I'm raising a new issue to use engine methods for all the reads, #246, and changing this method.

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check core

**Comment by JosefTaylor** [comment]
@BHoMBot check installer

**Comment by JosefTaylor** [comment]
@BHoMBot check versioning

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Review by peterjamesnugent — COMMENTED** [review]
LGTM - changes made as requested. It does need testing before approval though (by others).

**Review by mchaf — APPROVED** [review]
LGTM

**Comment by JosefTaylor** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**